src/clientiface.cpp

```diff
@@ -90,24 +91,30 @@ void RemoteClient::GetNextBlocks (
 		return;
 	}
 
+	// Camera position and direction
+	v3f camera_pos = sao->getEyePosition();
+	v3f camera_dir = v3f(0,0,1);
+	camera_dir.rotateYZBy(sao->getPitch());
+	camera_dir.rotateXZBy(sao->getYaw());
+
 	v3f playerpos = sao->getBasePosition();
 	const v3f &playerspeed = player->getSpeed();
-	v3f playerspeeddir(0,0,0);
-	if(playerspeed.getLength() > 1.0*BS)
-		playerspeeddir = playerspeed / playerspeed.getLength();
-	// Predict to next block
-	v3f playerpos_predicted = playerpos + playerspeeddir*MAP_BLOCKSIZE*BS;
+
+	// Calculate a general "adjustment value".
+	// The magnitude of this depends on the speed and the angle between
+	// the players velocity and the camera direction.
+	// Limit to 0.0f in case player moves backwards.
+	// When a player moves forward but looks down or backwards the adjustment will be 0.
+	// (note that playerspeed is not normalized, that way we save a few vector operations)
+	const f32 adjust = std::max(camera_dir.dotProduct(playerspeed), 0.0f) / BS;
+
+	// Predict where player will be soon. Just add the speed. The factor two tested well.
+	v3f playerpos_predicted = playerpos + playerspeed * 2.0f;
```

Comment:
Yes, this can be less than the previous MAP_BLOCKSIZE * BS, but that's OK here. This simply adjusts by speed.

Reply:
I can't see why this situation is different, might as well limit it to a minimum of 1 mapblock.
When walking at 4n/s the predicted point will be 8 nodes ahead, which half the time will not be in the mapblock ahead, so this would be a regression of prediction.

Reply:
This predicted point is now twice as far ahead as the first version of the PR, and also is no longer limited, so i would expect more issues with block islands (but have not tested yet). For example when falling at very high speed, like 200n/s, the predicted point will be 400 nodes below so the block island problem would be extreme.

It seems that the best method *might* be to keep predicted position only 1 or 2 mapblocks ahead, and improve performance other ways.